pkg/cluster/bootstrap.go

```diff
@@ -29,29 +34,157 @@ func (c *Cluster) Bootstrap(ctx context.Context) error {
 		return err
 	}
 
-	shouldBootstrap, err := c.shouldBootstrapLoad(ctx)
+	shouldBootstrap, isInitialized, err := c.shouldBootstrapLoad(ctx)
 	if err != nil {
 		return err
 	}
 	c.shouldBootstrap = shouldBootstrap
 
+	if c.managedDB != nil {
+		// In the case of etcd, if the database has been initialized, it doesn't
+		// need to be bootstrapped however we still need to check the database
+		// and reconcile the bootstrap data. Below we're starting a temporary
+		// instance of etcd in the event that etcd certificates are unavailable,
+		// reading the data, and comparing that to the data on disk, all the while
+		// starting normal etcd.
+		isHTTP := c.config.JoinURL != "" && c.config.Token != ""
+		if isInitialized && !isHTTP {
+			tmpDataDir := filepath.Join(c.config.DataDir, "db", "tmp-etcd")
+			os.RemoveAll(tmpDataDir)
+			if err := os.Mkdir(tmpDataDir, 0700); err != nil {
+				return err
+			}
+			etcdDataDir := etcd.DBDir(c.config)
+			if err := createTmpDataDir(etcdDataDir, tmpDataDir); err != nil {
+				return err
+			}
+
+			args := executor.ETCDConfig{
+				DataDir:           tmpDataDir,
+				ForceNewCluster:   true,
+				ListenClientURLs:  "http://127.0.0.1:2399",
+				Logger:            "zap",
+				HeartbeatInterval: 500,
+				ElectionTimeout:   5000,
+				LogOutputs:        []string{"stderr"},
+			}
+			configFile, err := args.ToConfigFile()
+			if err != nil {
+				return err
+			}
+			cfg, err := embed.ConfigFromFile(configFile)
+			if err != nil {
+				return err
+			}
+
+			etcd, err := embed.StartEtcd(cfg)
+			if err != nil {
+				return err
+			}
+			defer func() {
+				etcd.Close()
+				if err := os.RemoveAll(tmpDataDir); err != nil {
+					logrus.Warn("failed to remove etcd temp dir", err)
+				}
```

Comment:
should we defer this separately, up above where it's created? If the config fails to parse, or etcd fails to start, we'll leak the temp dir.

Reply:
Good catch. I moved that in testing. Resolved.